### PR TITLE
[Data] `iter_torch_batches` updates

### LINKED ISSUE #37632: [Data] Improve the  `iter_torch_batches()` docstrings with examples.
### Description

Users may experience some friction when converting from Torch DataLoader to Ray Data's `iter_torch_batches()`, due to differences in behavior and unclear documentation.

- Torch DataLoader always requires the user to move tensors to the GPU device, while Ray Data automatically loads tensors to the GPU if no collate function is provided.
- The input of the collate function is a list of items in Torch, but it is a dictionary of NumPy arrays in Ray Data.

**Improvement:**
Update the `iter_torch_batches` [API docstring](https://docs.ray.io/en/latest/data/api/doc/ray.data.Dataset.iter_torch_batches.html) by:
- Adding more details in the argument descriptions of `devices`, `collate_fn`.
- Provide a short code snippet for using `collate_fn`
  - showing the input batch data type(`Dict[str, np.array]`) and the tensor-to-cuda movement.
  - text tokenization and padding might be a good candidate.

### Link

_No response_

### PR DESCRIPTION
**Pull request by amogkam**
Adds more documentation to `iter_torch_batches` docstring.
Changes the default value of `device` parameter to `"auto"` to make the behavior of automatic device transfer more explicit.

Closes https://github.com/ray-project/ray/issues/37632

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by scottjlee — APPROVED** [review]

File: python/ray/data/iterator.py [hunk 1]
```diff
@@ -256,28 +255,43 @@ def iter_torch_batches(
...
+            >>> import numpy as np
+            >>> import ray
+            >>> def collate_fn(batch: Dict[str, np.ndarray]) -> Any:
+            ...    return torch.stack(
+            ...     [torch.as_tensor(array) for array in batch.values()],
+            ...     axis=1
+            ...    )
+            >>> iterator = ray.data.from_items([{"col_1": 1, "col_2": 2}]).iterator()
+            >>> for batch in iterator.iter_torch_batches(collate_fn=collate_fn):
+                    print(batch)
```
Comment (scottjlee) [inline python/ray/data/iterator.py:293]: should there be a `...` preceding this line?

File: python/ray/data/iterator.py [hunk 2]
```diff
@@ -256,28 +255,43 @@ def iter_torch_batches(
...
+            {'id': tensor([ 8,  9, 11,  7])}
+
+            Use the ``collate_fn`` to customize how the tensor batch is created.
+            >>> from typing import Any, Dict
+            >>> import torch
+            >>> import numpy as np
+            >>> import ray
+            >>> def collate_fn(batch: Dict[str, np.ndarray]) -> Any:
+            ...    return torch.stack(
+            ...     [torch.as_tensor(array) for array in batch.values()],
```
Comment (scottjlee) [inline python/ray/data/iterator.py:288]: the indenting looks a bit off here, does the docs code allow single space as the indent?

File: python/ray/data/iterator.py [hunk 3]
```diff
@@ -256,28 +255,42 @@ def iter_torch_batches(
...
+            Use the ``collate_fn`` to customize how the tensor batch is created.
+            >>> from typing import Any, Dict
+            >>> import torch
+            >>> import numpy as np
+            >>> import ray
+            >>> def collate_fn(batch: Dict[str, np.ndarray]) -> Any:
+            ...     return torch.stack(
+            ...     [torch.as_tensor(array) for array in batch.values()],
+            ...     axis=1)
+            >>> iterator = ray.data.from_items([{"col_1": 1, "col_2": 2}]).iterator()
```
Comment (woshiyyya) [inline python/ray/data/iterator.py:290]: Can we use more than 1 items here? So that we can apply `collate_fn` on a more realistic batch.

File: python/ray/data/iterator.py [hunk 4]
```diff
@@ -296,14 +309,17 @@ def iter_torch_batches(
...
+            collate_fn: A function to convert a Numpy batch to a PyTorch tensor batch.
+                When this parameter is specified, the user should manually handle the
+                host to device data transfer outside of collate_fn.
+                This is useful for further processing the data after it has been
+                batched. Potential use cases include collating along a dimension other
+                than the first, padding sequences of various lengths, or generally
+                handling batches of different length tensors. If not provided, the
+                default collate function is used which simply converts the batch of
+                numpy arrays to a batch of PyTorch tensors. This API is still
+                experimental and is subject to change. This parameter cannot be used in
```
Comment (woshiyyya) [inline python/ray/data/iterator.py:321]: Can we also mention that we move torch tensors to the appropriate device by default?

File: python/ray/data/iterator.py [hunk 5]
```diff
@@ -256,28 +255,42 @@ def iter_torch_batches(
...
-            Callable[[Union[np.ndarray, Dict[str, np.ndarray]]], Any]
-        ] = None,
+        collate_fn: Optional[Callable[[Dict[str, np.ndarray]], Any]] = None,
         drop_last: bool = False,
         local_shuffle_buffer_size: Optional[int] = None,
         local_shuffle_seed: Optional[int] = None,
         # Deprecated.
         prefetch_blocks: int = 0,
-    ) -> Iterator["TorchTensorBatchType"]:
+    ) -> Iterator[Dict[str, "torch.Tensor"]]:
```
Comment (woshiyyya) [inline python/ray/data/iterator.py:264]: Is the return batch type is `Any` when `collate_fn` is provided?  

Does`-> Iterator[Union[Dict[str, List], Any]]` make sense?

**Comment by woshiyyya** [comment]
Thank you @amogkam! Left some comments here.

File: python/ray/data/iterator.py (hunk 1 above)
Comment (amogkam) [reply python/ray/data/iterator.py]: updated

File: python/ray/data/iterator.py (hunk 2 above)
Comment (amogkam) [reply python/ray/data/iterator.py:288]: updated, should be better now

File: python/ray/data/iterator.py (hunk 3 above)
Comment (amogkam) [reply python/ray/data/iterator.py]: updated

File: python/ray/data/iterator.py (hunk 4 above)
Comment (amogkam) [reply python/ray/data/iterator.py:325]: I removed the automatic movement

File: python/ray/data/iterator.py (hunk 5 above)
Comment (amogkam) [reply python/ray/data/iterator.py]: good catch, updated

**Review by c21 — APPROVED** [review]
LGTM, thanks @amogkam.

**Review by woshiyyya — APPROVED** [review]

File: python/ray/data/dataset.py [hunk 6]
```diff
@@ -3212,31 +3215,48 @@ def iter_torch_batches(
...
             >>> import ray
-            >>> for batch in ray.data.range( # doctest: +SKIP
+            >>> for batch in ray.data.range(
             ...     12,
-            ... ).iter_torch_batches(batch_size=4):
-            ...     print(batch.shape) # doctest: +SKIP
-            torch.Size([4, 1])
-            torch.Size([4, 1])
-            torch.Size([4, 1])
+            ... ).iterator().iter_torch_batches(batch_size=4):
```
Comment (bveeramani) [inline python/ray/data/dataset.py:3236]:
```suggestion
            ... ).iter_torch_batches(batch_size=4):
```

File: python/ray/data/dataset.py [hunk 7]
```diff
@@ -3212,31 +3215,48 @@ def iter_torch_batches(
...
+            >>> import ray
+            >>> def collate_fn(batch: Dict[str, np.ndarray]) -> Any:
+            ...     return torch.stack(
+            ...         [torch.as_tensor(array) for array in batch.values()],
+            ...         axis=1
+            ...     )
+            >>> iterator = ray.data.from_items([
+            ...     {"col_1": 1, "col_2": 2},
+            ...     {"col_1": 3, "col_2": 4}]).iterator()
+            >>> for batch in iterator.iter_torch_batches(collate_fn=collate_fn):
```
Comment (bveeramani) [inline python/ray/data/dataset.py:3255]:
```suggestion
            >>> dataset = ray.data.from_items([
            ...     {"col_1": 1, "col_2": 2},
            ...     {"col_1": 3, "col_2": 4}])
            >>> for batch in dataset.iter_torch_batches(collate_fn=collate_fn):
```

File: python/ray/data/dataset.py [hunk 8]
```diff
@@ -3252,16 +3272,24 @@ def iter_torch_batches(
                 The final batch may include fewer than ``batch_size`` rows if
                 ``drop_last`` is ``False``. Defaults to 256.
             dtypes: The Torch dtype(s) for the created tensor(s); if ``None``, the dtype
-                is inferred from the tensor data.
-            device: The device on which the tensor should be placed; if ``None``, the
-                torch tensor is constructed on the CPU.
+                is inferred from the tensor data. This parameter cannot be used in
+                conjunction with ``collate_fn``.
```
Comment (bveeramani) [inline python/ray/data/dataset.py:3276]:
```suggestion
                is inferred from the tensor data. You can't use this parameter with 
                ``collate_fn``.
```

File: python/ray/data/dataset.py [hunk 9]
```diff
@@ -3252,16 +3272,24 @@ def iter_torch_batches(
...
+            device: The device on which the tensor should be placed. Defaults to
+                "auto" which moves the tensors to the appropriate device when the
+                Dataset is passed to Ray Train and ``collate_fn`` is not provided.
+                Otherwise, defaults to CPU. This parameter cannot be used in
+                conjunction with ``collate_fn``.
```
Comment (bveeramani) [inline python/ray/data/dataset.py:3281]:
```suggestion
            device: The device on which the tensor should be placed. Defaults to
                ``"auto"`` which moves the tensors to the appropriate device when the
                :class:`~ray.data.Dataset` is passed to Ray Train and ``collate_fn`` isn't provided.
                Otherwise, defaults to CPU. You can't use this parameter with ``collate_fn``.
```

File: python/ray/data/dataset.py [hunk 10]
```diff
@@ -3252,16 +3272,24 @@ def iter_torch_batches(
...
+                When this parameter is specified, the user should manually handle the
+                host to device data transfer outside of collate_fn.
+                This is useful for further processing the data after it has been
+                batched. Potential use cases include collating along a dimension other
+                than the first, padding sequences of various lengths, or generally
+                handling batches of different length tensors. If not provided, the
+                default collate function is used which simply converts the batch of
+                numpy arrays to a batch of PyTorch tensors. This API is still
+                experimental and is subject to change. This parameter cannot be used in
+                conjunction with ``dtypes`` or ``device``.
```
Comment (bveeramani) [inline python/ray/data/dataset.py:3292]:
```suggestion
                When this parameter is specified, the user should manually handle the
                host to device data transfer outside of ``collate_fn``.
                This is useful for further processing the data after it has been
                batched. Potential use cases include collating along a dimension other
                than the first, padding sequences of various lengths, or generally
                handling batches of different length tensors. If not provided, the
                default collate function is used which simply converts the batch of
                numpy arrays to a batch of PyTorch tensors. This API is still
                experimental and is subject to change. You can't use this parameter in
                conjunction with ``dtypes`` or ``device``.
```

File: python/ray/data/iterator.py [hunk 11]
```diff
@@ -255,29 +259,47 @@ def iter_torch_batches(
...
+    ) -> Iterator["TorchBatchType"]:
         """Return a batched iterator of Torch Tensors over the dataset.
 
-        This iterator will yield single-tensor batches if the underlying dataset
-        consists of a single column; otherwise, it will yield a dictionary of
-        column-tensors. If looking for more flexibility in the tensor conversion (e.g.
-        casting dtypes) or the batch format, try using `.iter_batches` directly.
+        This iterator will will yield a dictionary of column-tensors. If looking for
+        more flexibility in the tensor conversion (e.g. casting dtypes) or the batch
+        format, try using `.iter_batches` directly.
```
Comment (bveeramani) [inline python/ray/data/iterator.py:274]:
```suggestion
        This iterator yields a dictionary of column-tensors. If looking for
        more flexibility in the tensor conversion (e.g. casting dtypes) or the batch
        format, try using :meth:`.iter_batches` directly.
```

**Review by bveeramani — APPROVED** [review]
